### PR TITLE
Fixing base64

### PR DESCRIPTION
**Pull request by danielknobe**
base64 implementation is currently broken.

The following need to be done:

- [x] Write tests covering common cases
- [x] Implementation must be endianness independent
- [x] Implementation encode/decode of strings should work as expected

### CODE REVIEW

File: src/base64.cpp [hunk 1]
```diff
@@ -110,14 +110,23 @@ static_assert( length(translation_table) == 64, "error: need 64 characters in th
...
+
+	unsigned const index2 = bitmask & ((byte_array[1] << 2) | (byte_array[2] >> 6));
+	assert(index2 < 64);
+	writer[2] = translation_table[index2];
+
+	unsigned const index3 = bitmask & (byte_array[2]);
+	assert(index3 < 64);
+	writer[3] = translation_table[index3];
+
+	writer += 4;
```
Comment (ngc92) [inline src/base64.cpp:129]: shouldn't this also use `std::advance` to be consistent? Also, given that we use an iterator interface here, should we use `writer[i]`? Or actually consider this to be an `output_iterator` and perform iterator increments and writes?

File: src/base64.cpp [hunk 2]
```diff
@@ -169,14 +178,26 @@ std::string encode( const char* begin, const char* end, int newlines)
...
+	byte_array[0] = (value0 << 2);
+	byte_array[0] |= value1 >> 4;
+
+	byte_array[1] = value1 << 4;
+	byte_array[1] |= value2 >> 2;
+
+	byte_array[2] = value2 << 6;
+	byte_array[2] |= value3;
+
+	reader += 4;
```
Comment (ngc92) [inline src/base64.cpp:200]: again, also use `std::advance`?

**Review by ngc92 — CHANGES_REQUESTED** [review]
I think the 3-byte `encode` and `decode` functions should not be exposed in the header file.
Also, we should probably run the unit tests as part of our CI pipeline #61.
If we get it running in the Android simulator, we should also get a test for different endianess, right?

File: test/Base64Test.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    BOOST_CHECK( decode(encode(text)) == text );
+}
+
+BOOST_AUTO_TEST_CASE( encode_decode_three )
+{
+    std::vector<uint8_t> text = { '1', '2', '3' };
+    BOOST_CHECK( decode(encode(text)) == text );
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (ngc92) [inline test/Base64Test.cpp:61]: Add at least one test case where `text` contains something that is not a valid base64 character.

File: src/base64.cpp (hunk 1 above)
Comment (danielknobe) [reply src/base64.cpp]: I changed to `std::advance`. In my opinion using `writer[i]` is ok here.

File: src/base64.cpp (hunk 2 above)
Comment (danielknobe) [reply src/base64.cpp]: Changed.

File: src/base64.cpp (hunk 1 above)
Comment (ngc92) [reply src/base64.cpp]: ok

**Comment by danielknobe** [comment]
> I think the 3-byte `encode` and `decode` functions should not be exposed in the header file.
> Also, we should probably run the unit tests as part of our CI pipeline #61.
> If we get it running in the Android simulator, we should also get a test for different endianess, right?
> 
> Should we (silently) put -1 in the outputs, or rather throw an exception if any invalid input is detected for the decoding?

In the current implementation of this PR we ignore each invalid input character and carry on as if nothing bad had happened. What do you think about it?

File: test/Base64Test.cpp (hunk 3 above)
Comment (danielknobe) [reply test/Base64Test.cpp:74]: Done.

**Comment by danielknobe** [comment]
Replays are currently not working. Not yet checked what is the problem here.

File: src/base64.cpp [hunk 4]
```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
...
 std::vector<uint8_t> decode(const std::string& data )
 {
 	// pre-allocate buffer
 	const size_t dataSize = data.size();
-	std::vector<uint8_t> buffer( dataSize / 4 * 3 + 4 );
+
+	// check input size
+	if (dataSize % 4 != 0) {
+		// check if we have any
+		int const linefeeds = std::count_if(data.cbegin(), data.cend(), [](char ch){return ch == '\n'; });
```
Comment (ngc92) [inline src/base64.cpp:248]: couldn't this use `std::count` directly?

File: src/base64.cpp [hunk 5]
```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
...
+	if( ((value2 & (~bitmask)) != 0 ) ||
+	    ((value3 & (~bitmask)) != 0 ) )
+	{
+		throw std::runtime_error("invalid base64 character");
+	}
+	res = value2 << 6;
+	res |= value3;
+
+	return res;
+}
```
Comment (ngc92) [inline src/base64.cpp:228]: is this function still used anywhere? If not, should we delete it?

**Comment by ngc92** [comment]
I've added a few comments and suggested some more test cases. Can you rebase this on current master, so we can add the new tests to the CI?

**Comment by ngc92** [comment]
btw, the fact that we are defining our own `make_index_sq` here is one more argument (though not a strong one) in favour of #30

File: src/base64.cpp (hunk 4 above)
Comment (danielknobe) [reply src/base64.cpp:248]: Good idea. Changed to std::isspace. We still expect 4 chars without whitespace (block) but allow any amount of whitespace between blocks.

File: src/base64.cpp (hunk 5 above)
Comment (danielknobe) [reply src/base64.cpp:228]: Removed.

**Comment by ngc92** [comment]
what do you think about the other comments I left in the previous review? I think you've added the `continue`. If you disagree, or I'm just wrong, could you add a comment and mark as resolved?
The most serious one is line 246. This would fail if the input was e.g.  `AB\n\n`, right? The overall size is divisible by four, so we never even count the number of whitespaces.

**Comment by danielknobe** [comment]
I'm sorry. It can't take a look in your review for some strange reason.

**Comment by ngc92** [comment]
I'll just copy the comments then.

**Comment by ngc92** [comment]
I think I just had to press a "submit review" button still :)

File: src/base64.cpp [hunk 6]
```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
...
 	while(reader != data.cend())
 	{
-		// decode the valid group
-		if( is_valid(*reader) )
+		if( reader[0] == '\n') 
 		{
-			decode( iter, reader );
-		// correct fill bytes
-		} else if( *reader == '=' )
+			std::advance(reader, 1);
```
[the remark this answers is not in the document]
Comment (ngc92) [reply src/base64.cpp:254]: this is done

**Comment by danielknobe** [comment]
Yes. Now everything is all right :D.

File: src/base64.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielknobe) [reply src/base64.cpp:254]: Done already in previous commits.

File: src/base64.cpp [hunk 7]
```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
...
+	if (dataSize % 4 != 0) {
+		// check if we have any
+		int const linefeeds = std::count_if(data.cbegin(), data.cend(), [](char ch){return ch == '\n'; });
+		if ((dataSize - linefeeds) % 4 != 0)
+		{
+			throw std::runtime_error("data length must be multiple of 4");
+		}
+	}
+	
+	std::vector<uint8_t> buffer( dataSize / 4 * 3 );
```
[the remark this answers is not in the document]
Comment (danielknobe) [reply src/base64.cpp:246]: Done.

File: src/base64.cpp [hunk 8]
```diff
@@ -152,60 +163,128 @@ std::string encode( const char* begin, const char* end, int newlines)
...
+}
 
 std::vector<uint8_t> decode(const std::string& data )
 {
 	// pre-allocate buffer
 	const size_t dataSize = data.size();
-	std::vector<uint8_t> buffer( dataSize / 4 * 3 + 4 );
+
+	// check input size
+	if (dataSize % 4 != 0) {
```
[the remark this answers is not in the document]
Comment (danielknobe) [reply src/base64.cpp:237]: You are right.

File: test/Base64Test.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+BOOST_AUTO_TEST_CASE( decode_wrong_chars )
+{
+    std::string base64text = "M=T=I=z=";
+    BOOST_CHECK_THROW(decode(base64text), std::runtime_error);
+}
+
+BOOST_AUTO_TEST_CASE( decode_wrong_length )
+{
+    std::string base64text = "MTI";
```
[the remark this answers is not in the document]
Comment (danielknobe) [reply test/Base64Test.cpp:70]: Done.

File: test/Base64Test.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    BOOST_CHECK_THROW(decode(base64text), std::runtime_error);
+}
+
+BOOST_AUTO_TEST_CASE( encode_decode_multiline )
+{
+    std::vector<uint8_t> text = { '1', '2', '3', '4', '5', '6', '7', '8', '9' };
+    BOOST_CHECK( decode(encode(text, 8)) == text );
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
[the remark this answers is not in the document]
Comment (danielknobe) [reply test/Base64Test.cpp:80]: Done.

**Review by ngc92 — APPROVED** [review]
I think now everything looks good :)

**Pull request merged**